Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,620 @@
+// Sources/SwiftProtobuf/SwiftProtobufError.swift
+//
+// Copyright (c) 2024 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+
+/// A SwiftProtobuf specific error.
+///
+/// All errors have a high-level ``SwiftProtobufError/Code-swift.struct`` which identifies the domain
+/// of the error. For example, an issue when encoding a proto into binary data will result in a
+/// ``SwiftProtobufError/Code-swift.struct/binaryEncodingError`` error code.
+/// Errors also include a message describing what went wrong and how to remedy it (if applicable). The
+/// ``SwiftProtobufError/message`` is not static and may include dynamic information such as the
+/// type URL for a type that could not be decoded, for example.
+public struct SwiftProtobufError: Error, Equatable, @unchecked Sendable {
+    // Note: @unchecked because we use a backing class for storage.
+
+    private var storage: Storage
+    private mutating func ensureStorageIsUnique() {
+        if !isKnownUniquelyReferenced(&self.storage) {
+            self.storage = self.storage.copy()
+        }
+    }
+    
+    private final class Storage: Equatable {
+        var code: Code
```

Comment:
Since `code` is basically what used to be the different error types, we only have a domain level for errors now, and for someone to tell the difference between two errors, they would have to match the `message`?

That feels very fragile as correctly a typo in a message is a breaking change.

With the existing enum system, when parsing delimited data, we added the new error types so you can tell the difference between when the framing part of the stream is bad vs. when the message within a frame had an issue. This way a caller can decide if the error forces the invalidation of the whole stream or if they can continue reading messages just discarding the bad one. But in this new model, they are all decoding errors, so they would have to inspect the message strings themselves to tell what specific error it is.

Should there be a model where they can act on specific errors, or do we have to group them all now to different `code` value where all would be acted on in the same way?

Looking at the tests, maybe they can compare the errors to the accessors instead, should we not expose `message` so they can't really check it and only dump the error in `description`, but still use it for equality to avoid the fragile ness?

Reply:
The idea here is that `code` is unique per error type and we shouldn't really overload a single `code` with different `message`s. If that is correctly done then users are able to pattern match on the `code` which will give them back the exhaustive matching from before.

Reply:
I agree we shouldn't use the `message` for equality, I really only conformed this error to `Equatable` for tests, but I've removed this now. I am asserting that both the code and message match in tests though, because I think it doesn't hurt to make sure the message is what we want it to be, but users will be forced to match on the code instead. I've left `description` unchanged, meaning the `message` is used there. 

I do think exposing the `message` provides no real value and on the contrary it opens the door to users matching on it (even though they shouldn't) so I would agree to making it `internal` instead.

With regards to the current enum system allowing callers to differentiate between the whole framing in a stream being bad vs a single message: we can add a new `Code` case, maybe something like `binaryStreamDecodingError`. You could then differentiate between each scenario by catching either `binaryDecodingError` or `binaryStreamDecodingError` error codes. How does that sound?

Reply:
More codes to allow actually acting on the errors makes sense to me.